### ISSUE TITLE
Error with prop_compose! closure may outlive the current function, but it borrows

### ISSUE THREAD

**Issue by velvia** [AltSysrq/proptest#131]
Sorry I'm a newbie to Rust and tried using your wonderful crate to create an interesting strategy to generate a u64 with two parameters: number of bits, and chance that that number will be a zero.   I'm getting the error below.  Can I not use the parameters in the function body of prop_compose!?  Rust has closures right?  Why can't it capture the parameters here?  and maybe is there a better way to write this?  Thanks.

```
error[E0373]: closure may outlive the current function, but it borrows `zero_chance`, which is owned by the current function
   --> src/nibblepacking.rs:431:1
    |
431 | / prop_compose! {
432 | |     /// zero_chance: 0..1.0 chance of obtaining a zero
433 | |     fn arb_maybezero_nbits_u64(nbits: usize, zero_chance: f32)(n in any::<u64>()) -> u64 {
434 | |         // assert!(zero_chance < 1.0);
435 | |         let zero_limit = (zero_chance * std::u64::MAX as f32) as u64;
    | |                           ----------- `zero_chance` is borrowed here
436 | |         if n < zero_limit { 0 } else { n % (1 << nbits) }
437 | |     }
438 | | }
    | |_^ may outlive borrowed value `zero_chance`
    |
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
help: to force the closure to take ownership of `zero_chance` (and any other referenced variables), use the `move` keyword
    |
11  | strat , move | $ crate :: proptest_helper ! ( @ _WRAPPAT ( $ ( $ var ) , * ) ) | $
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0373]: closure may outlive the current function, but it borrows `nbits`, which is owned by the current function
   --> src/nibblepacking.rs:431:1
    |
431 | / prop_compose! {
432 | |     /// zero_chance: 0..1.0 chance of obtaining a zero
433 | |     fn arb_maybezero_nbits_u64(nbits: usize, zero_chance: f32)(n in any::<u64>()) -> u64 {
434 | |         // assert!(zero_chance < 1.0);
435 | |         let zero_limit = (zero_chance * std::u64::MAX as f32) as u64;
436 | |         if n < zero_limit { 0 } else { n % (1 << nbits) }
    | |                                                  ----- `nbits` is borrowed here
437 | |     }
438 | | }
    | |_^ may outlive borrowed value `nbits`
    |
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
help: to force the closure to take ownership of `nbits` (and any other referenced variables), use the `move` keyword
    |
11  | strat , move | $ crate :: proptest_helper ! ( @ _WRAPPAT ( $ ( $ var ) , * ) ) | $
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors
```

**Comment by AltSysrq** [comment]
The macro probably needs togenerate a `move` closure for this to work. I'll look into this for the next release.

In the mean time, you'll need to pass the arguments from one stage to the next manually.

```rust
fn arb_maybezero_nbits_u64
    (nbits: usize, zero_chance: f32)
    (n in any::<u64>(), nbits in Just(nbits), zero_chance in Just(zero_chance)) 
    -> u64 
{
    // Your body here
}
```

That said, I think your strategy here can be more easily expressed directly.

```rust
fn arb_maybezero_nbits_u64
    (nbits: usize, zero_chance: f32)
    (is_zero in prop::bool::weighted(zero_chance as f64),
     n in 0u64..(1 << nbits))
    -> u64
{
    if is_zero { 0 } else { n }
}
```

**Comment by velvia** [comment]
@AltSysrq thanks, what you have below is definitely much more elegant and works well.   Thanks for an awesome prop testing library.  It found a corner case for me already. 
Leaving this issue open for you to track any changes.

> That said, I think your strategy here can be more easily expressed directly.
> 
> ```rust
> fn arb_maybezero_nbits_u64
>     (nbits: usize, zero_chance: f32)
>     (is_zero in prop::bool::weighted(zero_chance as f64),
>      n in 0u64..(1 << nbits))
>     -> u64
> {
>     if is_zero { 0 } else { n }
> }
> ```

**Comment by AltSysrq** [comment]
This should be fixed in 0.9.2..
